### DISCUSSION TITLE
Huawei examples - Calculates the Power (W) being produced by the Solar Inverter(s) and consumed by the House

### DISCUSSION THREAD

**Discussion by ji-pi** [slipx06/sunsynk-power-flow-card#332] [Huawei / iStore]
Hello,

First of all, thank you very much for sharing your work!

Sensor start at ligne 763, there is a variable "inverters_input " but no sensor defined.
In the availability part, there are 2 times the sensor 'sensor.inverters_active_power'.

The result is for me not fair.

Could you help me ?

Many thanks

**Comment by slipx06** [comment]
@Roving-Ronin can you take a look?

**Reply by Roving-Ronin** [reply]
@jpgazerro 

Thanks for spotting that, and taking the time to flag it.  Be interested in any other feedback or suggestions you have for Huawei / iStore?
Below is the fixed up code, if you want to paste into your yaml (and if OK I'll do a push to SunSynk).

```
  # Calculates the Power (W) being produced by the Solar Inverter(s) and consumed by the House.
  - sensor:    
    - name: Inverters - Solar Power to House
      unique_id: inverters_solar_power_to_house
      device_class: power
      state_class: measurement
      unit_of_measurement: W
      state: >-
        {% set grid_power = states('sensor.power_meter_active_power')|int(default=0) %}
        {% set inverters_input = states('sensor.inverters_input_power')|int(default=0) %}
        {% set inverters_output = states('sensor.inverters_active_power')|int(default=0) %}
        {% set battery = states('sensor.batteries_charge_discharge_power')|int(default=0) %}
        {% if inverters_output > 0 %}                      
          {% if grid_power > 0 %}                  
            {% if battery > 0 %} {{ inverters_output - grid_power }}             
            {% else %} {{ inverters_output + battery - grid_power | abs }}
            {% endif %}       
            {% else %} 0
          {% endif %}         
        {% endif %}
      availability: >-
        {{ (states('sensor.power_meter_active_power') not in [ 'unknown', 'none', 'unavailable' ])
        or (states('sensor.inverters_input_power') not in [ 'unknown', 'none', 'unavailable' ])
        or (states('sensor.inverters_active_power') not in [ 'unknown', 'none', 'unavailable' ])
        or (states('sensor.batteries_charge_discharge_power') not in [ 'unknown', 'none', 'unavailable' ]) }}

```

Edit: Formula Updated.

**Reply by Roving-Ronin** [reply]
@jpgazerro 
Sorry stuffed up my first reply, I've updated the yaml code above, if you can test it. Also in the card yaml where you enter the entity to sensor mapping, could you add the line for pv_total back in, so assuming 4 PV inputs, it looks like this:

```
    pv1_power_186: sensor.inverter_1_pv_1_power
    pv2_power_187: sensor.inverter_1_pv_2_power
    pv3_power_188: sensor.inverter_2_pv_1_power
    pv4_power_189: sensor.inverter_2_pv_2_power
    pv_total: sensor.inverters_input_power
    environment_temp: sensor.<your_location>_temp
```

This will change it so instead of the box showing the total of the PV power being calculated by the SunSynk card logic, it will instead use the derived sensor that is calculated from inverter1 and inverter2 input power. This is more accurate and typically I see ~30W difference from this, to the Inverter(s) Active Power / aka the box above the inverter icon showing how much power is going out of the inverter(s).

Also to double check all the other power to / from flow sensors, do you want to paste this below into the Developer Tools --> Template screen, it should show a list of all those sensors. With a seperate browser window showing the SunSynk card, it makes it easier to look and work out if the formula / maths is correct for the sensors. The biggest issue I find is trying to work it out, test it and do so with there being no sun when you need it or the battery charging / grid importing when you need it the other way around.

```
Batteries - Power to Grid             {{ states('sensor.batteries_power_to_grid') }}
Batteries - Power to House            {{ states('sensor.batteries_power_to_house') }}
Batteries - Power to EV Charger       {{ states('sensor.batteries_power_to_ev_charger') }}    
Inverters - Solar Power to House      {{ states('sensor.inverters_solar_power_to_house') }}
Inverters - Solar Power to Batteries  {{ states('sensor.inverters_solar_power_to_batteries') }}
Inverters - Solar Power to EV Charger {{ states('sensor.inverters_solar_power_to_ev_charger') }}   
Inverters - Solar Power to Grid       {{ states('sensor.inverters_solar_power_to_grid') }}
Power Meter - Grid to Batteries       {{ states('sensor.power_meter_grid_to_batteries') }}
Power Meter - Grid to House           {{ states('sensor.power_meter_grid_to_house') }}
Power Meter - Grid to EV Charger      {{ states('sensor.power_meter_grid_to_ev_charger') }}   
```

Can I ask what your setup is ? i.e. 
- What model of inverter and how many (1 phase / 3 phase) ?
- DDSU / DTSU-666H meter installed
- Any LUNA Battery and if so, how many batteries of what size 5/10/15kWh each ?
- Any EV charger or Generator / other things ?

**Reply by ji-pi** [reply]
@Roving-Ronin 

Hello, thanks for your answer.

My setup :

- 1 inverter SUN2000-10KTL-M1 -> 1 string input 21 pannels (3 phases / master)
- 1 inverter SUN2000-6KTL-M1 > 1 string input 21 pannels (3 phases / slave)
- 1 LUNA2000 10 kWh connected on the master
- DTSU-666H meter
- Backup Box
- I have an AcThor 9S for my hot water (HA inject the solar surplus to the AcThor)

For first, I have test your new yaml code for the Solar Power to House consumption. Currently since there is not much sun, I charge my battery at the low rate at night, before 6:00 a.m. (Not a full charge, this depends on the weather, currently a manual estimate)

To test the code, I set an alarm at 5:00 a.m. this morning to see the behavior. 

When there is no sun and my battery charged, my first analysis sometimes gives a negative result (a few watts). I think it comes from the loss during the DC / AC conversion. To correct this and test, when the result is <0,  I set a value to 0. Because the difference comes from the grid and is counted in the Grid to House sensor's.

In order to have as many differents states as possible, I will do some tests over the next few days and analyze the results according to your request. 

I am telling you again.

**Reply by Roving-Ronin** [reply]
@jpgazerro 

Updating and still working on these.. I think they will need tweaking for when its daytime and PV is producing power also.

```
  # Calculates the Power (W) being Imported from the GRID and consumed by the EV Charger.
  - sensor:    
    - name: Power Meter - Grid to EV Charger
      unique_id: power_meter_grid_to_ev_charger
      device_class: power                                                        
      state_class: measurement                                                   
      unit_of_measurement: W                                                     
      state: >-
        {% set grid_power = states('sensor.power_meter_active_power')|int(default=0) %}     
        {% set inverters_output = states('sensor.inverters_active_power')|int(default=0) %}
        {% set ev_charger = states('sensor.pro3em_em2_active_power')|int(default=0) %}
        
        {% if inverters_output <= 0 and grid_power <= 0 %} {{ ev_charger }}
        {% else %} 0                                                                  
        {% endif %}         
      availability: >-
        {{ (states('sensor.power_meter_active_power') not in [ 'unavailable', 'none', 'unknown' ] )
          and (states('sensor.inverters_active_power') not in [ 'unavailable', 'none', 'unknown' ]) 
          and (states('sensor.pro3em_em2_active_power') not in [ 'unavailable', 'none', 'unknown' ]) }}
```


```
  # Calculates the Power (W) being consumed from the Battery(s) to the EV Charger.
  - sensor:    
    - name: Batteries - Power to EV Charger
      unique_id: batteries_power_to_ev_charger
      device_class: power                                                             
      state_class: measurement                                                        
      unit_of_measurement: W                                                          
      state: >-
        {% set battery = states('sensor.batteries_charge_discharge_power')|int(default=0) %}
        {% set ev_charger = states('sensor.ev_charger_active_power')|int(default=0) %}
        {% set batteries_to_house = states('sensor.batteries_power_to_house')|int(default=0) %}

        {% if battery < 0 %} {{ batteries_to_house +  ev_charger | int }}  
        {% else %} 0                                                          
        {% endif %}
      availability: >-
        {{ (states('sensor.batteries_charge_discharge_power') not in [ 'unknown', 'none', 'unavailable' ])
        or (states('sensor.ev_charger_active_power') not in [ 'unknown', 'none', 'unavailable' ])
        or (states('sensor.batteries_power_to_house') not in [ 'unknown', 'none', 'unavailable' ]) }}
```
These still need checking during the day, so if the inverter / PV is producing what is the output and how do we work out the formula for some power from PV, some maybe from Battery and some from Grid to then break it down to what amount is come from where.

Hard enough without my own LUNA battery or an EV, if only I there was a room with big lights and all this gear inside as a test lab, that I could flick the lights on/off to simulate day and night, and then simulate loads upon the grid / battery / ev etc. Without it we are at the whim of the day / night etc.


BTW as I try and work on these and test / validate / fix the logic, the 'as I go' changes go into this branch.  https://github.com/Roving-Ronin/sunsynk-power-flow-card/blob/development/docs/examples/huawei_packages/huawei_derived_sensors.yaml
